lib/libzpool/uzfs_io.c

```diff
@@ -36,29 +35,22 @@ uzfs_write_data(zvol_state_t *zv, char *buf, uint64_t offset, uint64_t len,
 	objset_t *os = zv->zv_objset;
 	rl_t *rl, *mrl;
 	int ret = 0, error;
-	uint64_t r_offset, r_len;
-	uint64_t r_moffset, r_mlen;
+	uint64_t r_offset;
 	metaobj_blk_offset_t metablk;
 	uint64_t metadatasize = zv->zv_volmetadatasize;
 	uint64_t len_in_first_aligned_block = 0;
 
-	sync = dmu_objset_syncprop(os);
+	sync = (dmu_objset_syncprop(os) == ZFS_SYNC_ALWAYS) ? 1 : 0;
 	if (zv->zv_volmetablocksize == 0)
 		metadata = NULL;
-	/*
-	 * Taking lock on entire block at ZFS layer.
-	 * Handling the case where readlen is smaller than blocksize.
-	 * This can also be avoided later for better performance.
-	 */
-	r_offset = (offset / blocksize) * blocksize;
-	r_len = ((len + blocksize - 1) / blocksize) * blocksize;
 
+	r_offset = P2ALIGN_TYPED(offset, blocksize, uint64_t);
```

Comment:
I don't see us handling a case when `offset` is already blocksize aligned. `r_offset` will be `offset`. len_in_first_aligned_block will be `blocksize` instead of zero.

Reply:
this is handled in the while loop @jkryl .. 
```        while (offset < end && offset < volsize) {
                if (len_in_first_aligned_block != 0) {
                        bytes = len_in_first_aligned_block;
                        len_in_first_aligned_block = 0;
                }
                else
                        bytes = (len < blocksize) ? len : blocksize;
```

Reply:
oh I see. thx